File: scripts/test_solutions/039_combination_sum.py
def combinationSum(candidates, target):
    result = []

    def backtrack(start, remaining, path):
        if remaining == 0:
            result.append(path[:])
            return
        if remaining < 0:
            return
        for i in range(start, len(candidates)):
            path.append(candidates[i])
            backtrack(i, remaining - candidates[i], path)
            path.pop()

    backtrack(0, target, [])
    return result

File: scripts/test_solutions/039_combination_sum.js
var combinationSum = function(candidates, target) {
    const result = [];
    function backtrack(start, remaining, path) {
        if (remaining === 0) { result.push([...path]); return; }
        if (remaining < 0) return;
        for (let i = start; i < candidates.length; i++) {
            path.push(candidates[i]);
            backtrack(i, remaining - candidates[i], path);
            path.pop();
        }
    }
    backtrack(0, target, []);
    return result;
};
